File: src/main/java/com/matteopierro/linkstation/domain/model/LinkStation.java
package com.matteopierro.linkstation.domain.model;

import static com.matteopierro.linkstation.domain.model.Power.ZERO;

public class LinkStation {
    private final Point point;
    private final Reach reach;

    public LinkStation(Point point, Reach reach) {
        this.point = point;
        this.reach = reach;
    }

    public Power powerFor(Device device) {
        var distanceFromDevice = device.distanceFrom(point);
        if (distanceFromDevice.isGreaterThan(reach)) {
            return ZERO;
        } else {
            return reach.minus(distanceFromDevice).square();
        }
    }

    @Override
    public String toString() {
        return "LinkStation at (" + point + ")";
    }
}

File: src/link-station.js
import Point from "./point";

export default class LinkStation {
    constructor(point, reach) {
        this.point = point;
        this.reach = reach;
    }

    powerFor(anotherPoint) {
        const distance = anotherPoint.distanceFrom(this.point);
        if (distance > this.reach) {
            return 0;
        } else {
            return Number(
                Math.pow(this.reach - distance, 2).toFixed(2)
            );
        }
    }
}

export const linkStation = (x, y, reach) => new LinkStation(new Point(x, y), reach);